### PR TITLE
improve syntax highlighting for certain languages (#1642)

### PR DESCRIPTION
**Pull request by jbhoosreddy**
Associated Issue: #1642 

### Summary of Changes

* Add support for JSX, Elm, CoffeeScript and TypeScript, Flow

### CODE REVIEW

**Comment by jasonLaster** [comment]
Just thinking about it. We can link to other language examples too. We
don't need to copy pasta everything

**Comment by jbhoosreddy** [comment]
I was thinking of adding examples in devtools-html/debugger-examples next.

File: bin/dev-server.js [hunk 1]
```diff
@@ -20,4 +20,4 @@ app.get("/integration/mocha.css", function(req, res) {
 });
 
 console.log("View debugger examples here:")
-console.log("https://github.com/jasonLaster/debugger-examples")
+console.log("https://github.com/devtools-html/debugger-examples")
```
Comment (clarkbw) [inline bin/dev-server.js]: 👍 I have been meaning to do this!

**Comment by jasonLaster** [comment]
Amazing!

We've got to put together beautiful screenshots before we can merge this. We need to market this!  💰

**Comment by jbhoosreddy** [comment]
@jasonLaster to test it out I need to figure out enabling source-maps. With that, I can try different examples. I might have to fire-up nightly with the right flags, maybe. @clarkbw. I'm open to suggestions.

**Comment by jasonLaster** [comment]
screen shot 2017-01-12 at 5 20 16 pm
screen shot 2017-01-12 at 5 19 46 pm

**Comment by jbhoosreddy** [comment]
@jasonLaster this should be good to go.

File: src/components/Editor.js [hunk 2]
```diff
@@ -320,6 +320,17 @@ const Editor = React.createClass({
...
+      this.editor.setMode({ name: "javascript", typescript: true });
+    } else if (contentType.includes("coffeescript")) {
+      this.editor.setMode("coffeescript");
+    } else if (contentType.includes("typescript-jsx")) {
+      this.editor.setMode({ name: "jsx",
+        base: { name: "javascript", typescript: true }});
+    } else if (contentType.includes("jsx")) {
+      this.editor.setMode("jsx");
+    } else if (contentType.includes("elm")) {
+      this.editor.setMode("elm");
```
Comment (jasonLaster) [inline src/components/Editor.js]: can we pull this into an object like 

```js
contentTypeModeMap = {
  typescript: { name: "javascript", typescript: true },
  coffeescript: "coffeescript"
}

function getMode(contentType) {
  // Object.keys(contentTypeModeMap)
 
}
```

```js
  this.editor.setMode(getMode(contentType))
```

File: src/utils/source-map-worker.js [hunk 3]
```diff
@@ -201,12 +201,27 @@ async function getOriginalSourceText(originalSource: Source) {
...
+  if (url.match(/ts|tsx/)) {
+    return "text/typescript";
+  }
+
+  if (url.match(/coffee$/)) {
+    return "text/x-coffeescript";
+  }
+
+  return "text/plain";
+}
```
Comment (jasonLaster) [inline src/utils/source-map-worker.js]: should we improve this now or wait for other cases?

File: src/utils/source-editor.js [hunk 4]
```diff
@@ -40,8 +53,10 @@ class SourceEditor {
     return this.editor.getValue();
   }
 
-  setMode(value) {
-    this.editor.setOption("mode", value);
+  setMode(contentType) {
```
Comment (jasonLaster) [inline src/utils/source-editor.js]: hmm - this sadly won't work due to the fact that there's a different panel source editor.

We'll need to use https://github.com/devtools-html/debugger.html/blob/master/src/utils/source.js

**Comment by jasonLaster** [comment]
Hey @jbhoosreddy, I made a couple tweaks. Mostly moving more logic to utils and adding some unit tests.

**Comment by jbhoosreddy** [comment]
@jasonLaster you did? I don't see them. +1 on the mocha test runner though! I'm trying it out. love it!

**Comment by jasonLaster** [comment]
@jbhoosreddy lol you're too quick. had a test error :)

File: src/utils/source-map-worker.js [hunk 5]
```diff
@@ -201,39 +201,10 @@ async function getOriginalSourceText(originalSource: Source) {
...
-  }
-
-  if (url.match(/elm/)) {
-    return "text/elm";
-  }
-
-  return "text/plain";
-}
-
 function applySourceMap(generatedId, url, code, mappings) {
```
Comment (jbhoosreddy) [inline src/utils/source-map-worker.js]: in retrospect, this is a nice idea. 👍

File: src/utils/source.js [hunk 6]
```diff
@@ -111,7 +111,8 @@ const contentTypeModeMap = {
 function getMode(sourceText: SourceText) {
   const { contentType, text } = sourceText;
 
-  if (text.match(/^\s*\/\/ @flow/)) {
+  // // @flow or /* @flow */
+  if (text.match(/^\s*(\/\/ @flow|\/\* @flow \*\/)/)) {
```
Comment (jbhoosreddy) [inline src/utils/source.js]: nice catch. I didn't know about this one.

**Comment by jasonLaster** [comment]
We'll need to add these modes to MC before this lands, but that shouldn't be too bad

**Comment by clarkbw** [comment]
🎉 👍 🎆

**Comment by jbhoosreddy** [comment]
☺️

**Pull request merged**